### PR TITLE
PIDController class was added to physics.control.

### LINKED ISSUE #26163: Create a PID controller object
There are probably many different specialized transfer functions that are often used. For example, a PID controller is a common transfer function. We could have an object that the user passes the gains and the transfer function is created.

```python
import sympy as sm
import sympy.physics.control as cn

KD, KP, KI = sm.symbols('K_D, K_P, K_I', real=True)
s = sm.symbols('s')

Pcont = cn.TransferFunction(KP, 1, s)
Icont = cn.TransferFunction(KI, s, s)
Dcont = cn.TransferFunction(KD*s, 1, s)

PIDcont = cn.Parallel(Pcont, Icont, Dcont)
```

```
In [3]: PIDcont
Out[3]: Parallel(TransferFunction(K_P, 1, s), TransferFunction(K_I, s, s), TransferFunction(K_D*s, 1, s))

In [4]: PIDcont.doit()
Out[4]: TransferFunction(K_I + s*(K_D*s + K_P), s, s)
```

### PR DESCRIPTION
**Pull request by abhiphile**
#### References to other Issues or PRs
Fixes: https://github.com/sympy/sympy/issues/26163

#### Brief description of what is fixed or changed
Introduced a PIDController class in the control module to simplify the creation and manipulation of PID controllers in control systems. This class allows users to define PID controllers by specifying proportional (kp), integral (ki), derivative (kd) gains and Derivative filter time constant (td).

#### Release Notes

* physics.control
  * Introduced the PIDController class in the control module.

### CODE REVIEW

File: sympy/physics/control/lti.py [hunk 1]
```diff
@@ -4263,3 +4263,134 @@ def is_controllable(self):
 
         """
         return self.controllability_matrix().rank() == self.num_states
+
+class PIDController(TransferFunction):
+    """
+    A PID Controller class that uses Parallel connection to combine the P, I, and D components.
+    Parameters
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4270]:
```suggestion
    A PID Controller class that uses Parallel connection to combine the P, I, and D components.
    
    Parameters
```

File: sympy/physics/control/lti.py [hunk 2]
```diff
@@ -4263,3 +4263,134 @@ def is_controllable(self):
...
+        Proportional gain.
+    KI : Expr, Number
+        Integral gain.
+    KD : Expr, Number
+        Derivative gain.
+    TP : Expr, Number
+        Derivative filter time constant
+    s : Symbol
+        The complex frequency variable.
+    Examples
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4282]:
```suggestion
        The complex frequency variable.
        
    Examples
```

File: sympy/physics/control/lti.py [hunk 3]
```diff
@@ -4263,3 +4263,134 @@ def is_controllable(self):
...
+    A PID Controller class that uses Parallel connection to combine the P, I, and D components.
+    Parameters
+    ==========
+    KP : Expr, Number
+        Proportional gain.
+    KI : Expr, Number
+        Integral gain.
+    KD : Expr, Number
+        Derivative gain.
+    TP : Expr, Number
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4278]: This is written as TF everywhere else so make it consistent. 
I was not aware of this being added to the implementation by the way. Why should it be considered ?

File: sympy/physics/control/lti.py [hunk 4]
```diff
@@ -4263,3 +4263,134 @@ def is_controllable(self):
...
+    >>> print(pid)
+    PIDController(1, 0.1, 0.01, 0, s)
+    >>> pid.doit() #Converts PIDController into TransferFunction
+    TransferFunction(s*(0.01*s + 1) + 0.1, s, s)
+    >>> pid.var
+    s
+    >>> pid.KP
+    1
+    """
+    def __new__(cls, KP, KI, KD, TF, var):
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4299]:
```suggestion
    def __new__(cls, KP=None, KI=None, KD=None, TF=None, var):
```
I think what you have done would work as well but let's give default values explicitly.

File: sympy/physics/control/lti.py (hunk 3 above)
Comment (abhiphile) [reply sympy/physics/control/lti.py:4278]: I've gone through the documentation of MATLAB for it https://in.mathworks.com/help/control/ug/proportional-integral-derivative-pid-controllers.html, and It included Tf.

File: sympy/physics/control/lti.py [hunk 5]
```diff
@@ -4263,3 +4263,134 @@ def is_controllable(self):
...
+
+        numerator = tf.num
+        denominator = tf.den
+
+        instance = super(PIDController, cls).__new__(cls,numerator,denominator,var)
+        instance._num = tf.num
+        instance._den = tf.den
+
+        return instance
+    def __init__(self, KP, KI, KD, TF, var):
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4336]:
```suggestion
        return instance
        
    def __init__(self, KP, KI, KD, TF, var):
```
Remember to leave lines after function definitions.

Comment (faze-geek) [reply sympy/physics/control/lti.py:4336]: And follow it everywhere in the PR.

File: sympy/physics/control/lti.py (hunk 4 above)
Comment (abhiphile) [reply sympy/physics/control/lti.py:4299]: Ohk sir, I'll give it in next commit.

**Comment by faze-geek** [comment]
This is a good start. I have left some reviews.

File: sympy/physics/control/lti.py (hunk 3 above)
Comment (abhiphile) [reply sympy/physics/control/lti.py:4278]: Should I remove the Tf from here ?

Comment (faze-geek) [reply sympy/physics/control/lti.py:4278]:
```suggestion
    TF : Expr, Number
```
No, I just wanted the argument name to be consistent with the code.

Comment (abhiphile) [reply sympy/physics/control/lti.py:4278]: That was my mistake, I've not noticed that thank for pointing out.

File: sympy/physics/control/lti.py [hunk 6]
```diff
@@ -4263,3 +4263,151 @@ def is_controllable(self):
...
+        numerator = tf.num
+        denominator = tf.den
+
+        instance = super(PIDController, cls).__new__(cls,numerator,denominator,var)
+        instance._num = tf.num
+        instance._den = tf.den
+
+        return instance
+
+    def __init__(self, KP, KI, KD, TF, var):
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4341]: Can't this initialization be done in the `__new__` constructor itself?

Comment (abhiphile) [reply sympy/physics/control/lti.py]: Yes, sir, it can be done by using `instance._TF = TF` here. 
I've used this in newer commit.

File: sympy/physics/control/lti.py [hunk 7]
```diff
@@ -4263,3 +4263,144 @@ def is_controllable(self):
...
+
+    @property
+    def KP(self):
+        """
+        Get the Proportional (KP) gain of the PIDController.
+
+        Returns
+        =======
+        Expr, Number
+            The Proportional gain.
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:4357]: Unnecessary as first line is self-explanatory. Follow for all class variables.

File: sympy/physics/control/tests/test_lti.py [hunk 8]
```diff
@@ -1681,6 +1681,37 @@ def test_StateSpace_functions():
...
+    assert p1.KD == kd
+
+    #Functionality Checking
+    assert p1.doit() == TransferFunction(ki + s*(kd*s + kp), s, s)
+    assert p1.to_expr().expand() == kd*s + ki/s + kp
+
+    #Using PIDController with TransferFunctions
+    tf1 = TransferFunction(s,1,s)
+    #Series combination and Parallel combination with TransferFunction
+    ser1 = Series(p1,tf1)
```
Comment (faze-geek) [inline sympy/physics/control/tests/test_lti.py:1707]:
```suggestion
    ser1 = Series(p1, tf1)
```

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1707]: Try to make the spacing organized, I see many missed spaces after commas which disturbs overall readability of code.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: I apologize, sir. I failed to notice it. In the future, I will make sure to thoroughly double-check everything before making any commits.

**Comment by faze-geek** [comment]
Hi @moorepants, is this ready ?

**Review by faze-geek — APPROVED** [review]
Looks fine to me, but lets wait for others to review the design and code.

File: sympy/physics/control/lti.py [hunk 9]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+    PIDController(1, 0.1, 0.01, 0, s)
+    >>> pid.doit() #Converts PIDController into TransferFunction
+    TransferFunction(s*(0.01*s + 1) + 0.1, s, s)
+    >>> pid.var
+    s
+    >>> pid.KP
+    1
+    """
+
+    def __new__(cls, KP = None, KI = None, KD = None, TF = None, var = None):
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4300]: The capital letters and spacing around equals does not follow PEP8.

File: sympy/physics/control/lti.py [hunk 10]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+    TF : Expr, Number
+        Derivative filter time constant
+    s : Symbol
+        The complex frequency variable.
+
+    Examples
+    ========
+    >>> from sympy.abc import s
+    >>> from sympy.physics.control.lti import PIDController
+    >>> KP, KI, KD, TF = 1, 0.1, 0.01, 0
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4288]: It is best if the examples show the symbolic use.

File: sympy/physics/control/lti.py [hunk 11]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
 
         """
         return self.controllability_matrix().rank() == self.num_states
+
+class PIDController(TransferFunction):
+    """
+    A PID Controller class that uses Parallel connection to combine the P, I, and D components.
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4269]: This needs more explanation. Show a blow diagram for the PID and what the "derivative filter time constant" is.

File: sympy/physics/control/lti.py [hunk 12]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+        """
+        Get the Derivative filter time constant (TP) of the PIDController.
+        """
+        return self._TF
+
+    def doit(self):
+        """
+        Convert the PIDController into a TransferFunction.
+        """
+        return TransferFunction(self._num, self._den, self.var)
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4381]: Why is this needed? If you subclass from TransferFunction, shouldn't `.doit()` work as expected if you properly define the numerator and denominators on initialization?

File: sympy/physics/control/tests/test_lti.py [hunk 13]
```diff
@@ -1679,3 +1679,34 @@ def test_StateSpace_functions():
     assert ss3.input_matrix == Matrix([[0, 0], [1, 0], [0, 1], [0, 0]])
     assert ss3.output_matrix == Matrix([[0, 1, 0, 0], [0, 0, 1, 0]])
     assert ss3.feedforward_matrix == Matrix([[0, 0], [0, 1]])
+
+
+def test_PIDControl():
+    kp, ki, kd = symbols(['kp', 'ki', 'kd'])
+    p1 = PIDController(kp, ki, kd,0, s)
```
Comment (moorepants) [inline sympy/physics/control/tests/test_lti.py:1686]:
```suggestion
    p1 = PIDController(kp, ki, kd, 0, s)
```

File: sympy/physics/control/lti.py (hunk 9 above)
Comment (moorepants) [reply sympy/physics/control/lti.py:4300]: Shouldn't var be a required argument?

File: sympy/physics/control/lti.py [hunk 14]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+
+    Examples
+    ========
+    >>> from sympy.abc import s
+    >>> from sympy.physics.control.lti import PIDController
+    >>> KP, KI, KD, TF = 1, 0.1, 0.01, 0
+    >>> pid = PIDController(KP, KI, KD, TF, s)
+    >>> print(pid)
+    PIDController(1, 0.1, 0.01, 0, s)
+    >>> pid.doit() #Converts PIDController into TransferFunction
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4292]:
```suggestion
    >>> pid.doit()
```

File: sympy/physics/control/lti.py [hunk 15]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+        Derivative gain.
+    TF : Expr, Number
+        Derivative filter time constant
+    s : Symbol
+        The complex frequency variable.
+
+    Examples
+    ========
+    >>> from sympy.abc import s
+    >>> from sympy.physics.control.lti import PIDController
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4287]:
```suggestion
    >>> from sympy.physics.control import PIDController
```

File: sympy/physics/control/lti.py [hunk 16]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
         """
         return self.controllability_matrix().rank() == self.num_states
+
+class PIDController(TransferFunction):
+    """
+    A PID Controller class that uses Parallel connection to combine the P, I, and D components.
+
+    Parameters
+    ==========
+    KP : Expr, Number
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4273]:
```suggestion
    KP : sympifiable
```

Comment (moorepants) [reply sympy/physics/control/lti.py:4273]: Same for all and you should have some tests to ensure that does work.

File: sympy/physics/control/lti.py [hunk 17]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+    >>> pid.KP
+    1
+    """
+
+    def __new__(cls, KP = None, KI = None, KD = None, TF = None, var = None):
+        if var is None:
+            raise ValueError("var should not be None")
+
+        if not isinstance(var, Symbol):
+            raise ValueError("var must be a Symbol")
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4305]: This checking should be inherited from transferfunction using `__super__()`.

File: sympy/physics/control/lti.py [hunk 18]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+            KD = S.Zero
+
+        KP = _sympify(KP)
+        KI = _sympify(KI)
+        KD = _sympify(KD)
+        TF = _sympify(TF)
+
+        P = TransferFunction(KP, 1, var)
+        I = TransferFunction(KI, var, var)
+        D = TransferFunction(KD * var, S.One+TF*var, var)
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4326]:
```suggestion
        D = TransferFunction(KD * var, S.One + TF*var, var)
```

PEP8

File: sympy/physics/control/lti.py [hunk 19]
```diff
@@ -4263,3 +4263,119 @@ def is_controllable(self):
...
+
+        pid_controller = Parallel(P, I, D)
+
+        tf = pid_controller.doit()
```
Comment (moorepants) [inline sympy/physics/control/lti.py:4330]: .doit() should not be called internally to any TransferFunction object.

**Comment by moorepants** [comment]
This is on the right track but again it doesn't follow the design principles we've established for TransferFunctions.

File: sympy/physics/control/lti.py (hunk 9 above)
Comment (faze-geek) [reply sympy/physics/control/lti.py:4300]: Yes, it doesn't need a default value and a check if it has been given.

Comment (faze-geek) [reply sympy/physics/control/lti.py:4300]: I think you did that because you cannot have a positional argument without a default value after a positional argument with a default value in a function in Python, if I'm correct ?

Comment (abhiphile) [reply sympy/physics/control/lti.py:4300]: Yes, It was showing error otherwise.
